File: lib/go/qdp.go
package qdp

import (
	"context"
	"encoding/binary"
	"errors"
	"io"
	"strings"
	"sync"
	"time"

	"github.com/rqure/qlib/pkg/log"
)

// Common errors
var (
	ErrInvalidMessage = errors.New("invalid message format")
	ErrCRCMismatch    = errors.New("CRC checksum mismatch")
)

// IConnectionHandler defines callbacks for transport connection status
type IConnectionHandler interface {
	OnConnect(transport ITransport)
	OnDisconnect(transport ITransport, err error)
}

// ConnectionHandlerFunc is a function type that implements IConnectionHandler
type ConnectionHandlerFunc struct {
	OnConnectFunc    func(transport ITransport)
	OnDisconnectFunc func(transport ITransport, err error)
}

func (h ConnectionHandlerFunc) OnConnect(transport ITransport) {
	if h.OnConnectFunc != nil {
		h.OnConnectFunc(transport)
	}
}

func (h ConnectionHandlerFunc) OnDisconnect(transport ITransport, err error) {
	if h.OnDisconnectFunc != nil {
		h.OnDisconnectFunc(transport, err)
	}
}

// ITransport defines the interface for different transport mechanisms
type ITransport interface {
	ReadMessage() (*Message, error)
	WriteMessage(*Message) error
	Close() error
}

// Message represents a QDP protocol message
type Message struct {
	Topic   string
	Payload []byte
}

// Encode serializes a Message into a byte slice with CRC
func (m *Message) Encode() ([]byte, error) {
	if m.Topic == "" {
		return nil, ErrInvalidMessage
	}

	topicLen := uint32(len(m.Topic))
	payloadLen := uint32(len(m.Payload))

	// Calculate total message size
	totalSize := 8 + // TopicLength + PayloadLength fields
		topicLen +
		payloadLen +
		4 // CRC32

	buf := make([]byte, totalSize)
	offset := 0

	// Write header
	binary.LittleEndian.PutUint32(buf[offset:], topicLen)
	offset += 4
	binary.LittleEndian.PutUint32(buf[offset:], payloadLen)
	offset += 4

	// Write topic
	copy(buf[offset:], m.Topic)
	offset += int(topicLen)

	// Write payload
	copy(buf[offset:], m.Payload)
	offset += int(payloadLen)

	// Calculate and write CRC
	crc := calculateCRC32(buf[:offset])
	binary.LittleEndian.PutUint32(buf[offset:], crc)

	return buf, nil
}

// Decode deserializes a byte slice into a Message, verifying CRC
func Decode(data []byte) (*Message, error) {
	if len(data) < 12 { // minimum size: 4 + 4 + 0 + 0 + 4 (lengths + CRC)
		return nil, ErrInvalidMessage
	}

	// Read header
	topicLen := binary.LittleEndian.Uint32(data[0:4])
	payloadLen := binary.LittleEndian.Uint32(data[4:8])

	totalLen := 8 + topicLen + payloadLen + 4
	if uint32(len(data)) != totalLen {
		return nil, ErrInvalidMessage
	}

	// Verify CRC
	receivedCRC := binary.LittleEndian.Uint32(data[totalLen-4:])
	calculatedCRC := calculateCRC32(data[:totalLen-4])
	if receivedCRC != calculatedCRC {
		return nil, ErrCRCMismatch
	}

	// Extract topic and payload
	topic := string(data[8 : 8+topicLen])
	payload := data[8+topicLen : totalLen-4]

	return &Message{
		Topic:   topic,
		Payload: payload,
	}, nil
}

// IMessageRxHandler defines callback interface for message handlers
type IMessageRxHandler interface {
	OnMessageRx(msg *Message)
}

// IMessageTxHandler defines callback interface for message handlers
type IMessageTxHandler interface {
	OnMessageTx(msg *Message)
}

// IMessageHandler defines the interface for message handlers
type IMessageHandler interface {
	IMessageRxHandler
	IMessageTxHandler
}

// MessageRxHandlerFunc is a function type that implements IMessageRxHandler
type MessageRxHandlerFunc func(msg *Message)

func (f MessageRxHandlerFunc) OnMessageRx(msg *Message) {
	f(msg)
}

// MessageTxHandlerFunc is a function type that implements IMessageTxHandler
type MessageTxHandlerFunc func(msg *Message)

func (f MessageTxHandlerFunc) OnMessageTx(msg *Message) {
	f(msg)
}

type MessageHandlerFunc struct {
	OnMessageRxFunc func(msg *Message)
	OnMessageTxFunc func(msg *Message)
}

func (f MessageHandlerFunc) OnMessageRx(msg *Message) {
	if f.OnMessageRxFunc != nil {
		f.OnMessageRxFunc(msg)
	}
}

func (f MessageHandlerFunc) OnMessageTx(msg *Message) {
	if f.OnMessageTxFunc != nil {
		f.OnMessageTxFunc(msg)
	}
}

// Subscription represents a topic subscription
type subscription struct {
	topic   string
	handler IMessageRxHandler
}

// ISubscriptionManager defines the interface for managing subscriptions
type ISubscriptionManager interface {
	Subscribe(topic string, handler IMessageRxHandler)
	Unsubscribe(topic string, handler IMessageRxHandler)
	OnMessageRx(msg *Message)
}

// DefaultSubscriptionManager implements ISubscriptionManager
type DefaultSubscriptionManager struct {
	subscriptions []subscription
	mu            sync.RWMutex
}

func NewSubscriptionManager() ISubscriptionManager {
	return &DefaultSubscriptionManager{
		subscriptions: make([]subscription, 0),
	}
}

func (sm *DefaultSubscriptionManager) Subscribe(topic string, handler IMessageRxHandler) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	sm.subscriptions = append(sm.subscriptions, subscription{topic, handler})
}

func (sm *DefaultSubscriptionManager) Unsubscribe(topic string, handler IMessageRxHandler) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	for i := len(sm.subscriptions) - 1; i >= 0; i-- {
		sub := sm.subscriptions[i]
		if sub.topic == topic && sub.handler == handler {
			sm.subscriptions = append(sm.subscriptions[:i], sm.subscriptions[i+1:]...)
		}
	}
}

func (sm *DefaultSubscriptionManager) OnMessageRx(msg *Message) {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	for _, sub := range sm.subscriptions {
		if matchTopic(sub.topic, msg.Topic) {
			sub.handler.OnMessageRx(msg)
		}
	}
}

// matchTopic checks if a topic matches a pattern with wildcards
func matchTopic(pattern, topic string) bool {
	if pattern == "#" {
		return true
	}

	patternParts := strings.Split(pattern, "/")
	topicParts := strings.Split(topic, "/")

	if len(patternParts) > len(topicParts) && patternParts[len(patternParts)-1] != "#" {
		return false
	}

	for i := 0; i < len(patternParts); i++ {
		if i >= len(topicParts) {
			return false
		}
		if patternParts[i] == "#" {
			return true
		}
		if patternParts[i] != "+" && patternParts[i] != topicParts[i] {
			return false
		}
	}

	return len(topicParts) == len(patternParts)
}

// IProtocol defines the interface for the QDP protocol implementation
type IProtocol interface {
	SendMessage(msg *Message) error
	ReceiveMessage() (*Message, error)
	Subscribe(topic string, handler IMessageRxHandler)
	Unsubscribe(topic string, handler IMessageRxHandler)
	PublishMessage(topic string, payload []byte) error
	StartReceiving(ctx context.Context)
	Close() error
}

// DefaultProtocol handles reading and writing messages over a transport
type DefaultProtocol struct {
	transport   ITransport
	subscribers ISubscriptionManager
	msgCh       chan *Message
	msgHandler  IMessageHandler
	ctx         context.Context
	cancel      context.CancelFunc
}

func NewProtocol(transport ITransport, subscribers ISubscriptionManager, msgHandler IMessageHandler) IProtocol {
	if subscribers == nil {
		subscribers = NewSubscriptionManager()
	}

	if msgHandler == nil {
		msgHandler = &MessageHandlerFunc{}
	}

	ctx, cancel := context.WithCancel(context.Background())
	return &DefaultProtocol{
		transport:   transport,
		subscribers: subscribers,
		msgHandler:  msgHandler,
		msgCh:       make(chan *Message, 100),
		ctx:         ctx,
		cancel:      cancel,
	}
}

// SendMessage sends a message over the transport
func (p *DefaultProtocol) SendMessage(msg *Message) error {
	err := p.transport.WriteMessage(msg)

	if err == nil {
		p.msgHandler.OnMessageTx(msg)
	}

	return err
}

// ReceiveMessage reads a message from the transport
func (p *DefaultProtocol) ReceiveMessage() (*Message, error) {
	return p.transport.ReadMessage()
}

// Subscribe adds a message handler for a topic pattern
func (p *DefaultProtocol) Subscribe(topic string, handler IMessageRxHandler) {
	p.subscribers.Subscribe(topic, handler)
}

// Unsubscribe removes a message handler for a topic pattern
func (p *DefaultProtocol) Unsubscribe(topic string, handler IMessageRxHandler) {
	p.subscribers.Unsubscribe(topic, handler)
}

// PublishMessage publishes a message to a topic
func (p *DefaultProtocol) PublishMessage(topic string, payload []byte) error {
	msg := &Message{
		Topic:   topic,
		Payload: payload,
	}
	return p.SendMessage(msg)
}

// handleMessage routes received messages to subscribers
func (p *DefaultProtocol) handleMessage(msg *Message) {
	p.msgHandler.OnMessageRx(msg)
	p.subscribers.OnMessageRx(msg)
}

// StartReceiving starts message processing loop
func (p *DefaultProtocol) StartReceiving(parentCtx context.Context) {
	ctx, cancel := context.WithCancel(parentCtx)

	// Start receive loop
	go func() {
		defer cancel()
		for {
			select {
			case <-ctx.Done():
				return
			default:
				msg, err := p.ReceiveMessage()
				if err != nil {
					select {
					case <-ctx.Done():
						return
					default:
						time.Sleep(100 * time.Millisecond)
						continue
					}
				}
				select {
				case p.msgCh <- msg:
				case <-ctx.Done():
					return
				}
			}
		}
	}()

	// Start process loop
	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			case msg := <-p.msgCh:
				p.handleMessage(msg)
			}
		}
	}()
}

// Close cancels the context and closes the transport
func (p *DefaultProtocol) Close() error {
	p.cancel() // Cancel our own context first
	return p.transport.Close()
}

var crc32Table [256]uint32

func init() {
	// Initialize CRC32 table
	polynomial := uint32(0xEDB88320)
	for i := uint32(0); i < 256; i++ {
		crc := i
		for j := 0; j < 8; j++ {
			if crc&1 != 0 {
				crc = (crc >> 1) ^ polynomial
			} else {
				crc >>= 1
			}
		}
		crc32Table[i] = crc
	}
}

func calculateCRC32(data []byte) uint32 {
	crc := uint32(0xFFFFFFFF)
	for _, b := range data {
		crc = (crc >> 8) ^ crc32Table[byte(crc)^b]
	}
	return crc ^ 0xFFFFFFFF
}

// Helper functions for message encoding/decoding
const (
	maxTopicSize   = 64
	maxMessageSize = 1024
)

// MessageReader handles buffered reading of QDP messages
type MessageReader struct {
	reader io.Reader
	buffer []byte
	pos    int
	count  int
}

// NewMessageReader creates a new buffered message reader
func NewMessageReader(reader io.Reader) *MessageReader {
	return &MessageReader{
		reader: reader,
		buffer: make([]byte, 2048), // Large enough for max message + potential garbage
	}
}

// ReadMessage now tries to process as many messages as possible from the buffer
func (r *MessageReader) ReadMessage() (*Message, error) {
	for {
		// Check if we need more data
		if r.count-r.pos < 8 {
			if err := r.readMore(); err != nil {
				return nil, err // Return immediately on read error
			}
			// If we still don't have enough data after reading
			if r.count-r.pos < 8 {
				continue
			}
		}

		// Look for valid message start
		for r.pos <= r.count-8 {
			topicLen := binary.LittleEndian.Uint32(r.buffer[r.pos:])
			payloadLen := binary.LittleEndian.Uint32(r.buffer[r.pos+4:])

			totalLen := 8 + topicLen + payloadLen + 4 // header + topic + payload + CRC

			// Validate message size
			if topicLen == 0 || topicLen > maxTopicSize ||
				payloadLen > maxMessageSize-8-topicLen-4 {
				r.pos++ // Invalid size, advance and continue scanning
				continue
			}

			// If we don't have enough data for the complete message
			if r.count-r.pos < int(totalLen) {
				// Try to read more data
				if err := r.readMore(); err != nil {
					return nil, err
				}
				// If we still don't have enough data, start over
				if r.count-r.pos < int(totalLen) {
					continue
				}
			}

			// Try to parse the message
			msg, err := r.tryParseMessage(int(totalLen))
			if err == nil {
				return msg, nil
			}
			// On parse error, advance one byte and continue scanning
			r.pos++
		}

		// If we've processed all data in the buffer, reset positions
		if r.pos >= r.count {
			r.pos = 0
			r.count = 0
		}

		// Read more data before continuing
		if err := r.readMore(); err != nil {
			return nil, err
		}
	}
}

func (r *MessageReader) readMore() error {
	// If there's unprocessed data, move it to start of buffer
	if r.pos < r.count {
		copy(r.buffer, r.buffer[r.pos:r.count])
		r.count -= r.pos
		r.pos = 0
	} else {
		r.count = 0
		r.pos = 0
	}

	// Check if buffer is full
	if r.count >= len(r.buffer) {
		return errors.New("buffer full")
	}

	// Read more data
	n, err := r.reader.Read(r.buffer[r.count:])
	if n > 0 {
		if n > 2 {
			log.Debug("Received %d bytes: % x", n, r.buffer[r.count:r.count+n])
		}
		r.count += n
	}
	if err == io.EOF && n > 0 {
		return nil // Return nil if we got some data even with EOF
	}
	return err
}

func (r *MessageReader) tryParseMessage(totalLen int) (*Message, error) {
	data := r.buffer[r.pos : r.pos+totalLen]

	// Verify CRC
	receivedCRC := binary.LittleEndian.Uint32(data[totalLen-4:])
	calculatedCRC := calculateCRC32(data[:totalLen-4])
	if receivedCRC != calculatedCRC {
		return nil, ErrCRCMismatch
	}

	// Extract message fields
	topicLen := binary.LittleEndian.Uint32(data[0:4])
	payloadLen := binary.LittleEndian.Uint32(data[4:8])
	topic := string(data[8 : 8+topicLen])
	payload := data[8+topicLen : 8+topicLen+payloadLen]

	// Advance past this message
	r.pos += totalLen

	return &Message{
		Topic:   topic,
		Payload: payload,
	}, nil
}

// Update existing readMessage to use MessageReader
func readMessage(r io.Reader) (*Message, error) {
	reader := NewMessageReader(r)
	return reader.ReadMessage()
}

func writeMessage(w io.Writer, msg *Message) error {
	// Calculate sizes
	topicLen := uint32(len(msg.Topic))
	payloadLen := uint32(len(msg.Payload))
	totalSize := 8 + topicLen + payloadLen + 4 // +4 for CRC

	// Create buffer for entire message
	msgData := make([]byte, totalSize)

	// Write header
	binary.LittleEndian.PutUint32(msgData[0:4], topicLen)
	binary.LittleEndian.PutUint32(msgData[4:8], payloadLen)

	// Write topic and payload
	copy(msgData[8:], msg.Topic)
	copy(msgData[8+topicLen:], msg.Payload)

	// Calculate and write CRC
	crc := calculateCRC32(msgData[:totalSize-4])
	binary.LittleEndian.PutUint32(msgData[totalSize-4:], crc)

	// Write entire message and log bytes
	n, err := w.Write(msgData)
	if n > 0 {
		log.Debug("Sent %d bytes: % x", n, msgData[:n])
	}
	return err
}

File: lib/c/qdp.c
#include "qdp.h"

#ifdef QDP_DEBUG
#define QDP_DEBUG_PRINT(...) fprintf(stderr, __VA_ARGS__)
#else
#define QDP_DEBUG_PRINT(...) ((void)0)
#endif

// CRC table and initialization
static uint32_t qdp_crc32_table[256];

static void qdp_crc32_init(void) {
    for (int i = 0; i < 256; i++) {
        uint32_t crc = i;
        for (int j = 0; j < 8; j++) {
            crc = (crc & 1) ? (crc >> 1) ^ 0xEDB88320 : crc >> 1;
        }
        qdp_crc32_table[i] = crc;
    }
}

static uint32_t qdp_calc_crc32(const uint8_t *data, size_t length) {
    uint32_t crc = 0xFFFFFFFF;
    for (size_t i = 0; i < length; i++) {
        crc = (crc >> 8) ^ qdp_crc32_table[(crc ^ data[i]) & 0xFF];
    }
    return crc ^ 0xFFFFFFFF;
}

// Add these helper functions for endian-safe operations
static void write_uint32_le(uint8_t* buf, uint32_t value) {
    buf[0] = (uint8_t)(value & 0xFF);
    buf[1] = (uint8_t)((value >> 8) & 0xFF);
    buf[2] = (uint8_t)((value >> 16) & 0xFF);
    buf[3] = (uint8_t)((value >> 24) & 0xFF);
}

static uint32_t read_uint32_le(const uint8_t* buf) {
    return ((uint32_t)buf[0]) |
           ((uint32_t)buf[1] << 8) |
           ((uint32_t)buf[2] << 16) |
           ((uint32_t)buf[3] << 24);
}

// Buffer operations
qdp_buffer_t qdp_buffer_create(uint8_t *data, size_t capacity) {
    return (qdp_buffer_t){
        .data = data,
        .size = 0,
        .capacity = capacity,
        .position = 0
    };
}

void qdp_buffer_reset(qdp_buffer_t *buf) {
    buf->size = 0;
    buf->position = 0;
}

bool qdp_buffer_can_read(const qdp_buffer_t *buf, size_t bytes) {
    return (buf->position + bytes) <= buf->size;
}

bool qdp_buffer_can_write(const qdp_buffer_t *buf, size_t bytes) {
    return (buf->size + bytes) <= buf->capacity;
}

// Message operations
bool qdp_message_write(qdp_buffer_t *buf, const qdp_message_t *msg) {
    // Calculate required size
    size_t header_size = 8;  // Just the lengths (4 + 4)
    size_t total_size = header_size + msg->header.topic_len + msg->payload.size + sizeof(uint32_t);
    
    QDP_DEBUG_PRINT("[QDP-WRITE] Writing message:\n");
    QDP_DEBUG_PRINT("  Topic: %s (len=%u)\n", msg->header.topic, msg->header.topic_len);
    QDP_DEBUG_PRINT("  Payload size: %u\n", msg->payload.size);
    QDP_DEBUG_PRINT("  Total size: %zu\n", total_size);
    
    if (!qdp_buffer_can_write(buf, total_size)) {
        QDP_DEBUG_PRINT("[QDP-WRITE] Buffer too small (capacity=%zu)\n", buf->capacity);
        return false;
    }

    size_t start_pos = buf->size;

    // Write lengths in little-endian
    write_uint32_le(buf->data + buf->size, msg->header.topic_len);
    buf->size += 4;
    write_uint32_le(buf->data + buf->size, msg->payload.size);
    buf->size += 4;

    // Write topic string
    memcpy(buf->data + buf->size, msg->header.topic, msg->header.topic_len);
    buf->size += msg->header.topic_len;

    // Write payload
    memcpy(buf->data + buf->size, msg->payload.data, msg->payload.size);
    buf->size += msg->payload.size;

    // Calculate CRC on everything before CRC position
    uint32_t crc = qdp_calc_crc32(buf->data + start_pos, buf->size - start_pos);
    write_uint32_le(buf->data + buf->size, crc);
    buf->size += sizeof(uint32_t);

    QDP_DEBUG_PRINT("[QDP-WRITE] Message written successfully (%zu bytes)\n", buf->size - start_pos);
    QDP_DEBUG_PRINT("[QDP-WRITE] CRC: %u\n", crc);
    
    return true;
}

bool qdp_message_read(qdp_buffer_t *buf, qdp_message_t *msg) {
    size_t start_pos = buf->position;
    
    QDP_DEBUG_PRINT("[QDP-READ] Attempting to read message at position %zu of %zu bytes\n", 
            buf->position, buf->size);

    // Dump first 16 bytes of buffer for debugging
    QDP_DEBUG_PRINT("[QDP-READ] Buffer head: ");
    for(size_t i = 0; i < 16 && (buf->position + i) < buf->size; i++) {
        QDP_DEBUG_PRINT("%02X ", buf->data[buf->position + i]);
    }
    QDP_DEBUG_PRINT("\n");

    // Read lengths (8 bytes total)
    if (!qdp_buffer_can_read(buf, 8)) {
        QDP_DEBUG_PRINT("[QDP-READ] Buffer too small for message header\n");
        return false;
    }

    // Read lengths in little-endian
    msg->header.topic_len = read_uint32_le(buf->data + buf->position);
    buf->position += 4;
    msg->header.payload_len = read_uint32_le(buf->data + buf->position);
    buf->position += 4;

    QDP_DEBUG_PRINT("[QDP-READ] Header read:\n");
    QDP_DEBUG_PRINT("  Topic length: %u\n", msg->header.topic_len);
    QDP_DEBUG_PRINT("  Payload length: %u\n", msg->header.payload_len);

    // Validate lengths
    if (msg->header.topic_len > QDP_MAX_TOPIC_LENGTH ||
        msg->header.payload_len > QDP_MAX_PAYLOAD_SIZE) {
        QDP_DEBUG_PRINT("[QDP-READ] Message exceeds max sizes (topic: %u/%u, payload: %u/%u)\n",
                msg->header.topic_len, QDP_MAX_TOPIC_LENGTH,
                msg->header.payload_len, QDP_MAX_PAYLOAD_SIZE);
        return false;
    }

    // Calculate remaining bytes from the original buffer size, not current position
    size_t total_message_size = 8 + msg->header.topic_len + msg->header.payload_len + sizeof(uint32_t);
    if (buf->size < total_message_size) {
        QDP_DEBUG_PRINT("[QDP-READ] Incomplete message (have %zu bytes, need %zu)\n",
                buf->size, total_message_size);
        return false;
    }

    // Read topic
    memcpy(msg->header.topic, buf->data + buf->position, msg->header.topic_len);
    msg->header.topic[msg->header.topic_len] = '\0';  // Null terminate
    buf->position += msg->header.topic_len;

    // Read payload
    memcpy(msg->payload.data, buf->data + buf->position, msg->header.payload_len);
    msg->payload.size = msg->header.payload_len;
    buf->position += msg->header.payload_len;

    // Read CRC in little-endian
    msg->checksum = read_uint32_le(buf->data + buf->position);
    buf->position += sizeof(uint32_t);

    // Verify CRC
    uint32_t calc_crc = qdp_calc_crc32(buf->data + start_pos, 
                                      buf->position - start_pos - sizeof(uint32_t));
    
    if (msg->checksum != calc_crc) {
        QDP_DEBUG_PRINT("[QDP] CRC mismatch - expected: %u, got: %u\n", msg->checksum, calc_crc);
        return false;
    }

    return true;
}

// Modified publish operations
int qdp_publish_string(qdp_t* qdp, const char* topic, const char* str) {
    qdp_stream_begin_message(&qdp->tx);
    
    qdp_message_t msg = {0};
    msg.header.topic_len = strlen(topic);
    msg.header.payload_len = strlen(str);
    strncpy(msg.header.topic, topic, QDP_MAX_TOPIC_LENGTH-1);
    
    size_t len = strlen(str);
    if (len > QDP_MAX_PAYLOAD_SIZE) return -1;
    memcpy(msg.payload.data, str, len);
    msg.payload.size = len;
    
    if (!qdp_message_write(&qdp->tx.buffer, &msg)) {
        return -1;
    }
    
    return qdp->transport.send ? qdp->transport.send(&qdp->tx.buffer, qdp->transport.ctx) : 0;
}

void qdp_message_clear(qdp_message_t *msg) {
    memset(&msg->header, 0, sizeof(qdp_header_t));
    memset(&msg->payload, 0, sizeof(qdp_payload_t));
    msg->checksum = 0;
}

uint32_t qdp_message_calc_size(const qdp_message_t *msg) {
    return sizeof(qdp_header_t) + msg->header.payload_len + sizeof(uint32_t);
}

// Stream operations
bool qdp_stream_begin_message(qdp_stream_t *stream) {
    qdp_buffer_reset(&stream->buffer);
    return true;
}

bool qdp_stream_end_message(qdp_stream_t *stream) {
    return stream->buffer.size > 0;
}

bool qdp_stream_next_message(qdp_stream_t *stream, qdp_message_t *msg) {
    // Don't try to read if we've reached the end
    if (stream->buffer.position >= stream->buffer.size) {
        return false;
    }
    return qdp_message_read(&stream->buffer, msg);
}

// Public API implementation
qdp_t* qdp_new(void) {
    qdp_t* qdp = QDP_MALLOC(sizeof(qdp_t));
    if (!qdp) return NULL;

    memset(qdp, 0, sizeof(qdp_t));
    qdp->rx.buffer = qdp_buffer_create(qdp->rx_data, QDP_MAX_BUFFER_CAPACITY);
    qdp->tx.buffer = qdp_buffer_create(qdp->tx_data, QDP_MAX_BUFFER_CAPACITY);
    
    qdp_crc32_init();
    return qdp;
}

void qdp_free(qdp_t* qdp) {
    if (qdp) QDP_FREE(qdp);
}

void qdp_set_transport(qdp_t* qdp, int (*send)(qdp_buffer_t*, void*),
                      int (*recv)(qdp_buffer_t*, void*), void* ctx) {
    qdp->transport.send = send;
    qdp->transport.recv = recv;
    qdp->transport.ctx = ctx;
}

bool qdp_subscribe(qdp_t* qdp, const char* pattern,
                  void (*callback)(const qdp_message_t*, void*), void* ctx) {
    if (qdp->sub_count >= QDP_MAX_SUBSCRIPTIONS) return false;
    
    qdp_subscription_t* sub = &qdp->subs[qdp->sub_count++];
    strncpy(sub->topic, pattern, QDP_MAX_TOPIC_LENGTH-1);
    sub->callback.fn = callback;
    sub->callback.ctx = ctx;
    return true;
}

void qdp_unsubscribe(qdp_t* qdp, const char* pattern) {
    for (uint32_t i = 0; i < qdp->sub_count; i++) {
        if (strcmp(qdp->subs[i].topic, pattern) == 0) {
            if (i < qdp->sub_count - 1) {
                memmove(&qdp->subs[i], &qdp->subs[i+1], 
                       sizeof(qdp_subscription_t) * (qdp->sub_count - i - 1));
            }
            qdp->sub_count--;
            i--; // Check same index again after shift
        }
    }
}

int qdp_publish_bytes(qdp_t* qdp, const char* topic, const void* data, size_t len) {
    if (len > QDP_MAX_PAYLOAD_SIZE) return -1;
    
    qdp_stream_begin_message(&qdp->tx);
    
    qdp_message_t msg = {0};
    msg.header.topic_len = strlen(topic);
    msg.header.payload_len = len;
    strncpy(msg.header.topic, topic, QDP_MAX_TOPIC_LENGTH-1);
    
    memcpy(msg.payload.data, data, len);
    msg.payload.size = len;
    
    if (!qdp_message_write(&qdp->tx.buffer, &msg)) {
        return -1;
    }
    
    return qdp->transport.send ? qdp->transport.send(&qdp->tx.buffer, qdp->transport.ctx) : 0;
}

int qdp_process(qdp_t* qdp) {
    if (!qdp->transport.recv) {
        QDP_DEBUG_PRINT("[QDP] No receive handler registered\n");
        return 0;
    }

    // Reset stream for new messages
    qdp_stream_begin_message(&qdp->rx);

    // Get new messages from transport
    int err = qdp->transport.recv(&qdp->rx.buffer, qdp->transport.ctx);
    if (err != 0) {
        QDP_DEBUG_PRINT("[QDP] Transport receive error: %d\n", err);
        return err;
    }

    if (qdp->rx.buffer.size == 0) {
        return 0;  // No data received
    }
    
    QDP_DEBUG_PRINT("[QDP] Received %zu bytes\n", qdp->rx.buffer.size);

    // Process messages using stream API
    qdp_message_t msg = {0};
    while (qdp_stream_next_message(&qdp->rx, &msg)) {
        QDP_DEBUG_PRINT("[QDP] Processing message: topic='%s' payload_len=%u\n", 
                msg.header.topic, msg.header.payload_len);
                
        // Match against subscriptions
        bool matched = false;
        for (uint32_t i = 0; i < qdp->sub_count; i++) {
            if (qdp_topic_matches(qdp->subs[i].topic, msg.header.topic)) {
                QDP_DEBUG_PRINT("[QDP] Message matched subscription '%s'\n", qdp->subs[i].topic);
                qdp->subs[i].callback.fn(&msg, qdp->subs[i].callback.ctx);
                matched = true;
            }
        }
        
        if (!matched) {
            QDP_DEBUG_PRINT("[QDP] No matching subscriptions for topic '%s'\n", msg.header.topic);
        }
        
        qdp_message_clear(&msg);
    }

    return 0;
}

// Helper functions
const char* qdp_get_topic(const qdp_message_t* msg) {
    return msg->header.topic;
}

const char* qdp_get_string(const qdp_message_t* msg) {
    if (msg->payload.size == 0) return NULL;
    return (const char*)msg->payload.data;
}

const void* qdp_get_bytes(const qdp_message_t* msg, size_t* len) {
    if (len) *len = msg->payload.size;
    return msg->payload.data;
}

bool qdp_topic_matches(const char* pattern, const char* topic) {
    const char *p = pattern;
    const char *t = topic;
    
    while (*p && *t) {
        // Handle multi-level wildcard
        if (*p == QDP_MULTI_LEVEL_WILDCARD) {
            return *(p + 1) == '\0';  // '#' must be last character
        }
        
        // Handle single-level wildcard
        if (*p == QDP_SINGLE_LEVEL_WILDCARD) {
            // Find next separator in topic
            const char* next_sep = strchr(t, QDP_TOPIC_LEVEL_SEPARATOR);
            if (!next_sep) {
                // No more separators, must match until end
                return *(p + 1) == '\0';
            }
            // Skip this level
            t = next_sep + 1;
            p++;
            if (*p == QDP_TOPIC_LEVEL_SEPARATOR) p++;
            continue;
        }
        
        // Handle separators
        if (*p == QDP_TOPIC_LEVEL_SEPARATOR) {
            if (*t != QDP_TOPIC_LEVEL_SEPARATOR) return false;
            p++;
            t++;
            continue;
        }
        
        // Match exact characters
        if (*p != *t) return false;
        p++;
        t++;
    }
    
    // Both strings must be at the end, or pattern ends with #
    return (*p == '\0' && *t == '\0') || 
           (*p == QDP_MULTI_LEVEL_WILDCARD && *(p + 1) == '\0');
}